library/core/src/slice/iter/macros.rs

```diff
@@ -191,6 +191,39 @@ macro_rules! iterator {
                 self.next_back()
             }
 
+            #[inline]
+            fn fold<B, F>(self, init: B, mut f: F) -> B
+                where
+                    F: FnMut(B, Self::Item) -> B,
+            {
+                // this implementation consists of the following optimizations compared to the
+                // default implementation:
+                // - do-while loop, as is llvm's preferred loop shape,
```

Comment:
It's very weird to me that doing this manually is needed, since LLVM has a loop rotation pass to do this.  But I guess it's fine.